cabal-install/Distribution/Client/Sandbox.hs

```diff
@@ -452,26 +455,47 @@ sandboxDeleteSource verbosity buildTreeRefs _sandboxFlags globalFlags = do
   (sandboxDir, pkgEnv) <- tryLoadSandboxConfig verbosity globalFlags
   indexFile            <- tryGetIndexFilePath (pkgEnvSavedConfig pkgEnv)
 
-  (removedPaths, convDict) <- Index.removeBuildTreeRefs verbosity indexFile buildTreeRefs
-  removeTimestamps sandboxDir removedPaths
+  (results, convDict) <- Index.removeBuildTreeRefs verbosity indexFile buildTreeRefs
 
-  let removedRefs = fmap (convertWith convDict) removedPaths
+  let (failedPaths, removedPaths) = partitionEithers results
+      removedRefs = fmap convDict removedPaths
+
+  unless (null removedPaths) $ do
+    removeTimestamps sandboxDir removedPaths
 
-  when (not . null $ removedPaths) $
     notice verbosity $ "Success deleting sources: " ++
-    showL removedRefs ++ "\n\n"
+      showL removedRefs ++ "\n\n"
 
-  when (length buildTreeRefs > length removedPaths) $
-    die $ "Skipped the following nonregistered sources: " ++
-    (showL $ buildTreeRefs \\ removedRefs)
+  unless (null failedPaths) $ do
+    mapM_ handleErrors $ groupBy errorType failedPaths
+    putNewlineStdErr
+    die $ "The sources with the above errors were skipped."
 
   notice verbosity $ "Note: 'sandbox delete-source' only unregisters the " ++
     "source dependency, but does not remove the package " ++
     "from the sandbox package DB.\n\n" ++
     "Use 'sandbox hc-pkg -- unregister' to do that."
   where
-    convertWith dict pth = fromMaybe pth $ fmap fst $ find ((==pth) . snd) dict
-    showL = concat . intersperse " " . fmap show
+    showPaths f = concat . intersperse " " . fmap (show . f)
+
+    putNewlineStdErr = hPutStr stderr "\n"
```

Comment:
But why do you need a newline before `Warning: ...`? We don't do that in other places in the code. Can you show an example of output with and without the newline?